pkg/gardenlet/bootstrap/certificate/certificate.go

```diff
@@ -0,0 +1,114 @@
+// Copyright (c) 2020 SAP SE or an SAP affiliate company. All rights reserved. This file is licensed under the Apache Software License, v. 2 except as noted otherwise in the LICENSE file
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package certificate
+
+import (
+	"context"
+	"crypto"
+	"crypto/x509/pkix"
+	"fmt"
+	"time"
+
+	bootstraputil "github.com/gardener/gardener/pkg/gardenlet/bootstrap/util"
+
+	"github.com/sirupsen/logrus"
+	certificatesv1beta1 "k8s.io/api/certificates/v1beta1"
+	certificatesv1beta1client "k8s.io/client-go/kubernetes/typed/certificates/v1beta1"
+	certutil "k8s.io/client-go/util/cert"
+	"k8s.io/client-go/util/certificate/csr"
+	"k8s.io/client-go/util/keyutil"
+)
+
+// RequestCertificate will create a certificate signing request for the Gardenlet
+// and send it to API server, then it will watch the object's
+// status, once approved by the gardener-controller-manager, it will return the kube-controller-manager's issued
+// certificate (pem-encoded). If there is any errors, or the watch timeouts, it
+// will return an error.
+func RequestCertificate(ctx context.Context, logger logrus.FieldLogger, certClient certificatesv1beta1client.CertificatesV1beta1Interface, seedName string, certificateSubject *pkix.Name) ([]byte, []byte, string, error) {
+	privateKeyData, err := keyutil.MakeEllipticPrivateKeyPEM()
+	if err != nil {
+		return nil, nil, "", fmt.Errorf("error generating client certificate private key: %v", err)
+	}
+
+	certData, csrName, err := requestSeedCertificate(ctx, logger, certClient.CertificateSigningRequests(), privateKeyData, seedName, certificateSubject)
+	if err != nil {
+		return nil, nil, "", err
+	}
+	return certData, privateKeyData, csrName, nil
+}
+
+func requestSeedCertificate(ctx context.Context, logger logrus.FieldLogger, certificateClient certificatesv1beta1client.CertificateSigningRequestInterface, privateKeyData []byte, seedName string, certificateSubject *pkix.Name) (certData []byte, csrName string, err error) {
+	if certificateSubject == nil {
+		certificateSubject = &pkix.Name{
+			Organization: []string{"gardener.cloud:system:seeds"},
+			CommonName:   "gardener.cloud:system:seed:" + seedName,
+		}
+	}
+	return requestCertificate(ctx, logger, certificateClient, privateKeyData, certificateSubject)
+}
```

Comment:
I suggest to remove this and let the bootstrapping logic pass the subject field instead.